File: packages/sdk-python/src/melaya/team.py
"""Team API — manage project team membership, roles, and invitations.

Maps to ``/api/v1/private/projects/:project/members/*`` and
``/api/v1/private/team/*``.

Example
-------
>>> from melaya import Melaya
>>> m = Melaya(api_key="mk_...")
>>> members = m.team.list_members("my-project")
>>> m.team.invite("my-project", username="alice")
>>> link = m.team.create_invite_link("my-project")
>>> m.team.accept_invite(token=link["token"])
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from .platform_types import JsonDict, TeamRole


class TeamAPI:
    def __init__(self, request: Any) -> None:
        self._request = request

    def list_members(self, project: str) -> List[JsonDict]:
        """List members of a project team."""
        return self._request("GET", f"/api/v1/private/projects/{project}/members")

    def invite(self, project: str, *, username: str) -> JsonDict:
        """Invite a user to a project team by username."""
        return self._request("POST", f"/api/v1/private/projects/{project}/members/invite",
                             json={"username": username})

    def create_invite_link(self, project: str) -> JsonDict:
        """Create a shareable invite link for a project."""
        return self._request("POST", f"/api/v1/private/projects/{project}/invite-link")

    def accept_invite(self, *, token: str) -> JsonDict:
        """Accept a project invite using the token from an invite link."""
        return self._request("POST", "/api/v1/private/team/invite/accept", json={"token": token})

    def update_member_role(self, project: str, user_id: str, *, role: TeamRole) -> JsonDict:
        """Update a team member's role in a project."""
        return self._request("PATCH", f"/api/v1/private/projects/{project}/members/{user_id}",
                             json={"role": role})

    def remove_member(self, project: str, user_id: str) -> JsonDict:
        """Remove a member from a project team."""
        return self._request("DELETE", f"/api/v1/private/projects/{project}/members/{user_id}")

    # ── Pipeline visibility ──────────────────────────────────────────────────────

    def get_pipeline_visibility(self, project: str, pipeline: str) -> JsonDict:
        """Get visibility settings for a pipeline within a project."""
        return self._request("GET", f"/api/v1/private/projects/{project}/pipelines/{pipeline}/visibility")

    def set_pipeline_visibility(self, project: str, pipeline: str, **body: Any) -> JsonDict:
        """Set pipeline visibility within a project."""
        return self._request("PUT", f"/api/v1/private/projects/{project}/pipelines/{pipeline}/visibility",
                             json=body)

File: packages/sdk-ruby/lib/melaya/team.rb
# frozen_string_literal: true

module Melaya
  # Team API — manage project team membership, roles, and invitations.
  #
  # Maps to:
  #   /api/v1/private/projects/:project/members/*
  #   /api/v1/private/team/*
  #   /api/v1/private/projects/:project/pipelines/:pipeline/visibility
  #
  # @example
  #   members = melaya.team.list_members("my-project")
  #   melaya.team.invite("my-project", username: "alice")
  #   link = melaya.team.create_invite_link("my-project")
  class TeamAPI
    def initialize(http)
      @http = http
    end

    # GET /api/v1/private/projects/:project/members
    # List members of a project team.
    # @param project [String] project name
    def list_members(project)
      @http.get("/api/v1/private/projects/#{enc(project)}/members")
    end

    # POST /api/v1/private/projects/:project/members/invite
    # Invite a user to a project team by username.
    # @param project [String]
    # @param username [String]
    def invite(project, username:)
      @http.post("/api/v1/private/projects/#{enc(project)}/members/invite",
        "username" => username)
    end

    # POST /api/v1/private/projects/:project/invite-link
    # Create a shareable invite link for a project.
    # Returns the URL to send to new team members.
    # @param project [String]
    def create_invite_link(project)
      @http.post("/api/v1/private/projects/#{enc(project)}/invite-link")
    end

    # POST /api/v1/private/team/invite/accept
    # Accept a project invite using the token from an invite link.
    # @param token [String]
    def accept_invite(token)
      @http.post("/api/v1/private/team/invite/accept", "token" => token)
    end

    # PATCH /api/v1/private/projects/:project/members/:userId
    # Update a team member's role in a project.
    # @param project [String]
    # @param user_id [String]
    # @param role [String] one of "owner", "editor", "viewer"
    def update_member_role(project, user_id, role:)
      @http.patch(
        "/api/v1/private/projects/#{enc(project)}/members/#{enc(user_id)}",
        "role" => role
      )
    end

    # DELETE /api/v1/private/projects/:project/members/:userId
    # Remove a member from a project team.
    # @param project [String]
    # @param user_id [String]
    def remove_member(project, user_id)
      @http.delete("/api/v1/private/projects/#{enc(project)}/members/#{enc(user_id)}")
    end

    # ── Pipeline visibility ────────────────────────────────────────────────────

    # GET /api/v1/private/projects/:project/pipelines/:pipeline/visibility
    # Get visibility settings for a pipeline within a project.
    # @param project [String]
    # @param pipeline [String]
    def get_pipeline_visibility(project, pipeline)
      @http.get(
        "/api/v1/private/projects/#{enc(project)}/pipelines/#{enc(pipeline)}/visibility"
      )
    end

    # PUT /api/v1/private/projects/:project/pipelines/:pipeline/visibility
    # Set pipeline visibility within a project.
    # @param project [String]
    # @param pipeline [String]
    # @param body [Hash]
    def set_pipeline_visibility(project, pipeline, body = {})
      @http.put(
        "/api/v1/private/projects/#{enc(project)}/pipelines/#{enc(pipeline)}/visibility",
        body
      )
    end

    private

    def enc(s)
      URI.encode_www_form_component(s.to_s)
    end
  end
end

